compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -307,55 +370,70 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         // relying on projections in the impl-trait-ref.
         //
         // e.g., `impl<U: Tr, V: Iterator<Item=U>> Foo<<U as Tr>::T> for V`
-        impl_obligations.append(&mut substs.obligations);
+        substs.obligations.append(&mut impl_obligations);
```

Comment:
What does interchanging these two do? Does it affect the logic or is it just a performance optimization?

Reply:
This affects semantics. Evaluation is done in a single pass. The change is so that evaluating `substs.obligations` resolves some type variables before evaluating `impl_obligations`.